pyxem/utils/cuda_utils.py

```diff
@@ -0,0 +1,133 @@
+    ----------
+    array : numpy or cupy array
+        Array to determine whether numpy or cupy should be used
+    Returns
+    -------
+    array : numpy.ndarray
+    """
+    if is_cupy_array(array):
+        import cupy as cp
+
+        array = cp.asnumpy(array)
+    return array
+
+
+def get_array_module(array):
+    """
+    Returns the array module for the given array
+    Parameters
+    ----------
+    array : numpy or cupy array
+        Array to determine whether numpy or cupy should be used
+    Returns
+    -------
+    module : module
+    """
+    module = np
+    try:
+        import cupy as cp
+
+        if isinstance(array, cp.ndarray):
+            module = cp
+    except ImportError:
+        pass
+    return module
+
+
+def is_cupy_array(array):
+    """
+    Convenience function to determine if an array is a cupy array
+
+    Parameters
+    ----------
+    array : array
+        The array to determine whether it is a cupy array or not.
+    Returns
+    -------
+    bool
+        True if it is cupy array, False otherwise.
+    """
+    try:
+        import cupy as cp
+
+        return isinstance(array, cp.ndarray)
+    except ImportError:
+        return False
+
+
+@cuda.jit
+def _correlate_polar_image_to_library_gpu(
+    polar_image, sim_r, sim_t, sim_i, correlation
+):
+    """
+    Custom cuda kernel for calculating the correlation for each template
+    in a library at all in-plane angles with a polar image
+
+    Parameters
+    ----------
+    polar_image: 2D numpy.ndarray or cupy.ndarray
+        The image in polar coordinates of shape (theta, R)
+    sim_r: 2D numpy.ndarray or cupy.ndarray
+        The r coordinates of all spots in all templates in the library.
+        Spot number is the column, template is the row
+    sim_t: 2D numpy.ndarray or cupy.ndarray
+        The theta coordinates of all spots in all templates in the library.
+        Spot number is the column, template is the row
+    sim_i: 2D numpy.ndarray or cupy.ndarray
+        The intensity of all spots in all templates in the library.
+        Spot number is the column, template is the row
+    correlation: 2D numpy.ndarray or cupy.ndarray
+        The output correlation matrix of shape (template number, theta), giving
+        the correlation of each template at each in-plane angle
+    """
+    # each cuda thread handles one element
+    template, shift = cuda.grid(2)
+    # don't calculate for grid positions outside
+    if template >= correlation.shape[0] or shift >= correlation.shape[1]:
+        return
+    tmp = 0.0
+    # add up all contributions to the correlation from spots
+    for spot in range(sim_r.shape[1]):
+        if sim_r[template, spot] == 0:
+            break
+        tmp += (
+            polar_image[
+                (sim_t[template, spot] + shift) % polar_image.shape[0],
+                sim_r[template, spot],
+            ]
+            * sim_i[template, spot]
+        )
+    correlation[template, shift] = tmp
```

Comment:
this looks very weird to me, could you explain?

Reply:
The advantage of a GPU is doing as much as possible in parallel on separate threads, since there are thousands of individual cores. For some image (converted to polar coordinates) we want to know the best matching template and the best-matching in-plane angle of said template. To find the optimal solution, we have to calculate the correlation for each template at each in-plane angle; this can all be done independently and in parallel. So we define a 2D "grid" of GPU threads, each one responsible for one template and one in-plane angle. This function is a "kernel" that is run on each of these threads in parallel. A thread knows about it's own "position" in the grid, which is information we can use to put the result in the correct place of the output array. With the line
```python
template, shift = cuda.grid(2)
```
each thread will know what position it is in in the grid. I will then map the first coordinate onto the template index and the second coordinate onto the "shift" (= in plane rotation) coordinate in the output array `correlation`. 

 Typically we want a 1-to-1 correspondence between the grid of threads and the output array. However, due to the architecture of the GPU, it may be that the grid is slightly larger and that some threads should do nothing (certainly not update any values). To ensure a thread doesn't do anything if it's outside the boundaries of the 2D array we want to calculate there's the lines:
```python
if template >= correlation.shape[0] or shift >= correlation.shape[1]:
        return
```
Then we set up the actual calculation each thread must perform. A temporary variable is declared which will be our result. 
```
tmp = 0.0
```
The value is stored in a local bit of memory for each thread. Each thread must then loop over all the diffraction spots in the template it is responsible for.  
```python
for spot in range(sim_r.shape[1]):
```
We loop over the the maximum number of spots but break the loop once it encounters a spot where `r=0` (not all templates have an equal number of spots, so we have padded with 0's) . 
```python
if sim_r[template, spot] == 0:
            break
```
Then to actually calculate the contribution  of each spot to the correlation we have:
```python
tmp += (
            polar_image[
                (sim_t[template, spot] + shift) % polar_image.shape[0],
                sim_r[template, spot],
            ]
            * sim_i[template, spot]
        )
```
which basically says: get me the pixel in the polar image that corresponds to the coordinate of the right spot, shifted along the theta direction by `shift` (modulo 360), and multiply it by the template spot intensity. This is basically our dot product.

When we have looped over all the spots in the template, we can place our calculated value in the right place of the output array.

After running the kernel over all shifts and templates (which happens in parallel) we find the maximum correlation by taking the maximum over the shift axis (happens elsewhere). 

Does this explain it adequately?

One could also do this simultaneously and instead of a 2D grid, have each thread be responsible for one template - a 1D grid, and calculating the optimum shift during the kernel's execution. However unintuitive, my benchmarks showed that this is actually slower that calculating the entire array of possibilities and then calculating the maximum element in each row (basically a bit map-reduce like).

I've thought of additional ideas to further optimize the entire procedure but for now I think this works well enough.

Reply:
Especially the explanation of the use of `cupy.grid()` was clarifying for me, thanks!

Reply:
That helps a lot, it's just very strange to see a "function" that doesn't end with a `return` - but as far as I understand it that's very standard for this sort of CUDA coding?

Reply:
Yep, cuda kernels can not allocate global memory as far as I understand, so you have to pass it a reference to the output array which should already be pre-allocated somehow. I made a little tutorial on cuda programming in python in case you're interested: https://youtu.be/9bBsvpg-Xlk

Reply:
Great, thanks for this